src/ringbuffer.h

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_RINGBUFFER_H
+#define BITCOIN_RINGBUFFER_H
+
+#include <array>
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+/**
+ * A threadsafe circular buffer.
+ *
+ * Allows optional blocking behavior instead of overwriting when at capacity.
+ */
+template<typename ET, int S, typename ST=int>
+class RingBuffer
+{
+public:
+    using value_type = ET;
+    using size_type = ST;
+
+    RingBuffer() : m_size(0), m_insert_at_idx(0), m_read_at_idx(0), m_stop_waiting(false) { }
+    ~RingBuffer() { }
+
+    /** @return the number of elements in the buffer. */
+    inline size_type Size() const {
+        std::lock_guard<std::mutex> lock(m_lock);
+        return m_size;
+    }
+    inline size_type MaxSize() const { return S; }
+
+    value_type PopFront()
+    {
+        std::lock_guard<std::mutex> lock(m_lock);
+        return AssumeLockedPopFront();
+    }
+
+    /**
+     * Push an element into the buffer.
+     *
+     * @arg no_overwrite  If true, block and wait to write instead of overwriting existing data.
+     */
+    size_type PushBack(value_type v, bool no_overwrite = true)
```

Comment:
I think having a move version of PushBack with const v& ref push back would make more sense to avoid unnecessary copying.
`PushBack(value_type&& v)` and `PushBack(const value_type& v)`

Reply:
Good catch. Shouldn't `const value_type&` bind to both rvalues and lvalues, though?

Reply:
Fixed.